sys/net/application_layer/unicoap/utils.c

```diff
@@ -0,0 +1,465 @@
+        if (res < 0) {
+            UNICOAP_DEBUG("buf too small " _NEED_HAVE "\n", iolist_size(message->payload_chunks), capacity);
+            return res;
+        }
+        message->payload_representation = UNICOAP_PAYLOAD_CONTIGUOUS;
+        message->payload = buffer;
+        message->payload_size = res;
+        return res;
+    }
+    case UNICOAP_PAYLOAD_CONTIGUOUS:
+        return message->payload_size;
+    default:
+        UNREACHABLE();
+        return -1;
+    }
+}
+
+void unicoap_message_payload_append_chunk(unicoap_message_t* message, iolist_t* chunk)
+{
+    assert(message->payload_representation == UNICOAP_PAYLOAD_NONCONTIGUOUS);
+    if (message->payload_chunks) {
+        iolist_append(&message->payload_chunks, chunk);
+    }
+    else {
+        message->payload_chunks = chunk;
+    }
+}
+
+static inline iolist_t* _append_payload_to_iolist(const unicoap_message_t* message, iolist_t* element)
+{
+    switch (message->payload_representation) {
+    case UNICOAP_PAYLOAD_NONCONTIGUOUS:
+        return message->payload_chunks;
+    case UNICOAP_PAYLOAD_CONTIGUOUS:
+        iolist_init(element, message->payload, message->payload_size, NULL);
+        return element;
+    default:
+        UNREACHABLE();
+        return NULL;
+    }
+}
+
+/**
+ * @pre 4 iolists
+ */
+int unicoap_pdu_buildv_options_and_payload(uint8_t* header, size_t header_size, const unicoap_message_t* message,
+                                           iolist_t* iolists)
+{
+    assert(header);
+    assert(message);
+    assert(iolists);
+    iolist_init(iolists, header, header_size, NULL);
+
+    iolist_t* element = iolists;
+    if (message->options && message->options->option_count > 0) {
+        element->iol_next = element + 1;
+        element += 1;
+        iolist_init(element, unicoap_message_options_data(message),
+                    unicoap_message_options_size(message), NULL);
+    }
+
+    if (unicoap_message_payload_get_size(message) > 0) {
+        element->iol_next = element + 1;
+        element += 1;
+
+        header[header_size] = UNICOAP_PAYLOAD_MARKER;
+        iolist_init(element, &header[header_size], 1, element + 1);
+
+        element->iol_next = _append_payload_to_iolist(message, element + 1);
+    }
+
+    return 0;
+}
+
+ssize_t unicoap_pdu_build_options_and_payload(uint8_t* pdu, size_t capacity, const unicoap_message_t* message)
+{
+    assert(pdu);
+    assert(message);
+
+    if (message->options && message->options->option_count > 0) {
+        if (capacity < unicoap_message_options_size(message)) {
+            return -ENOBUFS;
+        }
+        memcpy(pdu, unicoap_message_options_data(message), unicoap_message_options_size(message));
+        pdu += unicoap_message_options_size(message);
+        capacity -= unicoap_message_options_size(message);
+    }
+
+    if (unicoap_message_payload_get_size(message) > 0) {
+        *pdu = UNICOAP_PAYLOAD_MARKER;
+        pdu += 1;
+        capacity -= 1;
+
+        ssize_t size = 0;
```

Comment:
```suggestion
        ssize_t payload_size = 0;
```

slightly more expressive